File: algorithm/search/src/binary_search.py
# Dongyoung Kwon @Chuncheonian (ehddud2468@gmail.com)
# python version requirement: >=3.9

def binary_search(nums: list[int], target: int) -> int:
    lhs = 0
    rhs = len(nums) - 1

    while lhs <= rhs:
        mid = lhs + (rhs - lhs) // 2

        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            lhs = mid + 1
        else:
            rhs = mid - 1
    
    return -1

# Testcase
print(binary_search([1,2,3,4,5], 4))  # > 3

File: algorithm/search/src/binarySearch.swift
// Dongyoung Kwon @Chuncheonian (ehddud2468@gmail.com)

func binarySearch(nums: [Int], target: Int) -> Int {
    var (lhs, rhs) = (0, nums.count - 1)
        
    while lhs <= rhs {
        let mid = lhs + (rhs - lhs) / 2
            
        if nums[mid] == target {
            return mid
        }
        nums[mid] > target ? (rhs = mid - 1) : (lhs = mid + 1)
    }
    return -1
}

// Testcase
print(binarySearch(nums: [1, 2, 3, 4, 5], target: 4))  // > 3